src/schema.rs

```diff
@@ -46,35 +45,135 @@ impl Schema {
     /// otherwise returns None.
-    fn get_type<A: AsRef<str>>(&self, name: A) -> Option<&Type> {
+    fn get_type<A: AsRef<str>>(&self, name: A) -> Option<&TypeDefinition> {
         self.types.get_type_by_name(name.as_ref())
     }
 
     /// Returns an iterator over the types in this schema.
     pub(crate) fn get_types(&self) -> SchemaTypeIterator {
-        SchemaTypeIterator {
-            types: self.types.get_types().iter(),
-        }
+        SchemaTypeIterator::new(Rc::clone(&self.types))
     }
 
     /// Returns a new [Schema] instance containing all the types of this
     /// instance plus the provided type.  Note that the added type
     /// in the returned instance will hide a type of the same name
     /// from this instance.
-    fn plus_type(&self, schema_type: Type) -> Self {
+    fn plus_type(&self, schema_type: TypeDefinition) -> Self {
         todo!()
     }
 }
 
 /// Provides an Iterator which returns [Type]s inside a [Schema]
-pub struct SchemaTypeIterator<'a> {
-    types: Iter<'a, Type>,
+pub struct SchemaTypeIterator {
+    types: Rc<TypeStore>,
+    index: usize,
 }
 
-impl<'a> Iterator for SchemaTypeIterator<'a> {
-    type Item = &'a Type;
+impl SchemaTypeIterator {
+    fn new(types: Rc<TypeStore>) -> Self {
+        Self { types, index: 0 }
+    }
+}
+
+impl Iterator for SchemaTypeIterator {
+    type Item = TypeRef;
 
     fn next(&mut self) -> Option<Self::Item> {
-        self.types.next()
+        if self.index >= self.types.get_types().len() {
+            return None;
+        }
+        self.index = self.index + 1;
+        Some(TypeRef::new(self.index - 1, Rc::clone(&self.types)))
+    }
+}
+
+#[cfg(test)]
+mod schema_tests {
+    use super::*;
+    use crate::system::Resolver;
+    use ion_rs::value::owned::OwnedElement;
+    use ion_rs::value::reader::element_reader;
+    use ion_rs::value::reader::ElementReader;
+    use rstest::*;
+    use std::cell::RefCell;
+
+    // helper function to be used by isl tests
+    fn load(text: &str) -> Vec<OwnedElement> {
+        element_reader()
+            .read_all(text.as_bytes())
+            .expect("parsing failed unexpectedly")
+    }
+
+    #[rstest(
+    owned_elements, total_types,
+    case::type_constraint_with_anonymous_type(
+        /* For a schema with single anonymous type as below: 
+            type:: { type: int }
+         */
+        load(r#" // For a schema with single anonymous type
+            type:: {type: int}
+        "#).into_iter(),
+        2 // this includes the core type int and the anonymous type
+    ),
+    case::type_constraint_with_named_type(
+        load(r#" For a schema with named type as below: 
+            type:: { name: my_int, type: int }
+        "#).into_iter(),
+        2 // this includes the core type int and the anonymous type
+    ),
+    case::type_constraint_with_self_reference_type(
+        load(r#" For a schema with self reference type as below: 
+            type:: { name: my_int, type: my_int }
```

Comment:
Thought/nitpick—I've just realized that this particular case should actually be invalid ISL. It's logically impossible to resolve a type that defines itself in terms of `not`, `type`, `one_of`, `all_of`, or `any_of` itself. Self-referencing only makes sense in the scenarios where a thing is composed of other instances of the same type (eg. `element`, `ordered_element`, `fields`).

We can make a new issue to follow up on this later.

Reply:
I have opened an issue #27 for this which could be handled once the implementation of `element`, `fields`, `ordered_elements` is available.